cpp/src/table/row_operators.cu

```diff
@@ -77,50 +162,75 @@ auto decompose_structs(table_view table,
-    if (is_nested(col.type())) {
+  for (size_t col_idx = 0; col_idx < linked_columns.size(); ++col_idx) {
+    auto const& col = linked_columns[col_idx];
+    if (is_nested(col->type())) {
       // convert and insert
       std::vector<column_view> r_verticalized_columns;
       std::vector<int> r_verticalized_col_depths;
-      std::vector<column_view> flattened;
+      std::vector<LinkedColPtr> flattened;
       std::vector<int> depths;
       // TODO: Here I added a bogus leaf column at the beginning to help in the while loop below.
       //       Refactor the while loop so that it can handle the last case.
-      flattened.push_back(make_empty_column(type_id::INT32)->view());
-      std::function<void(column_view const&, int)> recursive_child = [&](column_view const& c,
-                                                                         int depth) {
+      flattened.push_back(
+        std::make_shared<linked_column_view>(make_empty_column(type_id::INT32)->view()));
+      std::function<void(LinkedColPtr, int)> recursive_child = [&](LinkedColPtr c, int depth) {
         flattened.push_back(c);
         depths.push_back(depth);
-        if (c.type().id() == type_id::STRUCT) {
-          for (int child_idx = 0; child_idx < c.num_children(); ++child_idx) {
-            auto scol = structs_column_view(c);
-            recursive_child(scol.get_sliced_child(child_idx), depth + 1);
+        if (c->type().id() == type_id::LIST) {
+          recursive_child(c->children[lists_column_view::child_column_index], depth + 1);
+        } else if (c->type().id() == type_id::STRUCT) {
+          for (auto& child : c->children) {
+            recursive_child(child, depth + 1);
           }
         }
       };
       recursive_child(col, 0);
       int curr_col_idx     = flattened.size() - 1;
-      column_view curr_col = flattened[curr_col_idx];
+      column_view curr_col = *flattened[curr_col_idx];
       while (curr_col_idx > 0) {
         auto const& prev_col = flattened[curr_col_idx - 1];
-        if (not is_nested(prev_col.type())) {
+        if (not is_nested(prev_col->type())) {
           // We hit a column that's a leaf so seal this hierarchy
+          // But first, traverse upward and include any list columns in the ancestors
+          linked_column_view* parent = flattened[curr_col_idx]->parent;
+          while (parent) {
+            if (parent->type().id() == type_id::LIST) {
+              // Include this parent
+              curr_col = column_view(
+                parent->type(),
+                parent->size(),
+                nullptr,  // list has no data of its own
+                nullptr,  // If we're going through this then nullmask is already in another branch
+                UNKNOWN_NULL_COUNT,
+                parent->offset(),
+                {parent->child(lists_column_view::offsets_column_index), curr_col});
+            }
+            parent = parent->parent;
+          }
```

Comment:
```suggestion
          for (linked_column_view* parent = flattened[curr_col_idx]->parent; parent; parent = parent->parent) {
            if (parent->type().id() == type_id::LIST) {
              // Include this parent
              curr_col = column_view(
                parent->type(),
                parent->size(),
                nullptr,  // list has no data of its own
                nullptr,  // If we're going through this then nullmask is already in another branch
                UNKNOWN_NULL_COUNT,
                parent->offset(),
                {parent->child(lists_column_view::offsets_column_index), curr_col});
            }
          }
```

I'm not sure if this should be further refactored, but I like this being in a for loop better because I think it makes it more clear what is going on here and you can't accidentally forget to change parent later, etc.